policy-controller/k8s/index/src/outbound/index.rs

```diff
@@ -370,76 +412,57 @@ impl Namespace {
                 };
+
                 let (opaque, accrual) = match service_info.get(&service_ref) {
                     Some(svc) => (svc.opaque_ports.contains(&sp.port), svc.accrual),
                     None => (false, None),
                 };
 
-                // The HttpRoutes which target this Service but don't specify
-                // a port apply to all ports. Therefore we include them.
-                let routes = self
-                    .service_routes
-                    .get(&sp.service)
-                    .cloned()
-                    .unwrap_or_default();
+                // The routes which target this Service but don't specify
+                // a port apply to all ports. Therefore, we include them.
+                let routes = self.service_routes.get(&sp.service).cloned();
 
                 let mut service_routes = ServiceRoutes {
                     opaque,
                     accrual,
                     authority,
-                    namespace: self.namespace.clone(),
-                    name: sp.service,
                     port: sp.port,
+                    name: sp.service,
+                    namespace: self.namespace.clone(),
                     watches_by_ns: Default::default(),
                 };
 
-                // Producer routes are routes in the same namespace as their
-                // parent service. Consumer routes are routes in other
-                // namespaces.
-                let (producer_routes, consumer_routes): (Vec<_>, Vec<_>) = routes
-                    .into_iter()
-                    .partition(|(gknn, _route)| *gknn.namespace == *self.namespace);
-                for (gknn, route) in consumer_routes {
-                    // Consumer routes should only apply to watches from the
-                    // consumer namespace.
-                    let watch = service_routes.watch_for_ns_or_default(gknn.namespace.to_string());
-                    watch.routes.insert(gknn, route);
-                }
-                for (gknn, route) in producer_routes {
-                    // Insert the route into the producer namespace.
-                    let watch = service_routes.watch_for_ns_or_default(gknn.namespace.to_string());
-                    watch.routes.insert(gknn.clone(), route.clone());
-                    // Producer routes apply to clients in all namespaces, so
-                    // apply it to watches for all other namespaces too.
-                    for (ns, watch) in service_routes.watches_by_ns.iter_mut() {
-                        if ns != &gknn.namespace {
-                            watch.routes.insert(gknn.clone(), route.clone());
-                        }
+                match routes {
+                    None => {}
+                    Some(OutboundRouteCollection::Http(routes)) => {
+                        service_routes
+                            .insert_producer_and_consumer_routes(routes, self.namespace.as_str());
```

Comment:
Code duplication reduction. It doesn't happen here (yet) because there's only one variant of `OutboundRouteCollection`, but once the `Grpc` variant is introduced all the code that's now inside `insert_producer_and_consumer_routes` would be duplicated otherwise.